### ISSUE TITLE
controlclient: support going online when control is unreachable (cache netmaps)

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#12639]
If the control plane is unreachable (either due to the local network conditions or a control server outage), it should still be possible to go online, at least in a limited state, by using a cached copy of the last known network configuration.

Most servers and subnet routers don't change their endpoints/DERP home often, so would still be reachable.

In the future we will also likely more control plane messaging to be between peers (as we did earlier with endpoint info) so the effect of a control outage would be even less. (e.g. imagine designating 1+ always-on tailscaled in your tailnet to be backup control plane messaging nodes that can take the role of the control plane during an outage. with the appropriate crypto it can all be trustless even, similar to the derp protocol)

In any case, first step is making the client cache netmaps so they can transition from offline to online (WantRunning false to true) and get to ipn.State Running without a control connection active. (a health warnable would still be firing saying the control connection was down)

/cc @andrew-d (who's started on this) @maisem @sailorfrag @raggi @agottardo

**Comment by dolceAlka** [comment]
Would this entail adding STUN to tailscaled?

**Comment by reesericci** [comment]
It seems I'm running into this issue, where `tailscale netcheck` outputs the following on the locked-down network I'm attempting to connect Tailscale on:

```
Error: CLI credentials are not available ErrorFromBackend("Unavailable")
```

For context, I'm able to connect to Mullvad by using UDP over TCP with port 80.

**Comment by ruifung** [comment]
I had a random thought when reading this issue, could tailscaled be able to use an available exit-node to establish connectivity to the control plane if it can establish connection to an exit-node using the cached netmap feature being worked on here?

This was just a thought because if you had a exit-node that has non-expiring keys and stable derp region, it could be a viable means of potentially reaching the controlplane to sync the latest state.

**Comment by b-** [comment]
> I had a random thought when reading this issue, could tailscaled be able to use an available exit-node to establish connectivity to the control plane if it can establish connection to an exit-node using the cached netmap feature being worked on here?

I think this feature is more relevant to the scenario of the control plane going down while you were already connected, than the scenario of not being able to access the control server (but being able to access an exit node that _can_ access the control server).

If you're trying to connect to Tailscale from a restricted network that blocks connections to the control server, can I suggest https://github.com/jaxxstorm/proxyt as a solution? It's a proxy that you run on your own server to relay a connection from a Tailscale client to the control server. It solves the problem of, "I cannot connect to the Tailscale control plane, but I can connect to my other machine."

**Comment by ruifung** [comment]
> > I had a random thought when reading this issue, could tailscaled be able to use an available exit-node to establish connectivity to the control plane if it can establish connection to an exit-node using the cached netmap feature being worked on here?
> 
> I think this feature is more relevant to the scenario of the control plane going down while you were already connected, than the scenario of not being able to access the control server (but being able to access an exit node that _can_ access the control server).
> 
> If you're trying to connect to Tailscale from a restricted network that blocks connections to the control server, can I suggest https://github.com/jaxxstorm/proxyt as a solution? It's a proxy that you run on your own server to relay a connection from a Tailscale client to the control server. It solves the problem of, "I cannot connect to the Tailscale control plane, but I can connect to my other machine."

Nah, this was just a random thought I had, since the issue mentions "local network conditions" is that if tailscaled can manage to reach an exit node, it could potentially route the control plane connection through it and update it's netmap/derpmap/policy. (also because the issue mentions potential future work on p2p control plane messaging between peers)

I don't think proxyt will let an existing client just switch controlplane servers for existing accounts, you'd end up having to log in as a new device.

Actually, now that I think about it, what would happen if I configured the tailscale control plane to go through an app connector...
